### PR TITLE
fn: allow setting CUDA stream

### PR DESCRIPTION
**Pull request by havogt**
The `fn::gpu` backend now supports setting the CUDA stream in which the kernel is launched.

### CODE REVIEW

File: tests/regression/fn/fn_tridiagonal_solve.cpp [hunk 1]
```diff
@@ -96,7 +96,10 @@ namespace {
         using float_t = typename TypeParam::float_t;
 
         auto fencil = [&](auto sizes, auto const &a, auto const &b, auto const &c, auto const &d, auto &x) {
-            auto be = fn_backend_t();
+            cudaStream_t s;
```
Comment (havogt) [inline tests/regression/fn/fn_tridiagonal_solve.cpp:99]: remove

File: include/gridtools/common/cuda_util.hpp [hunk 2]
```diff
@@ -74,14 +74,34 @@ namespace gridtools {
             return res;
         }
 
+// #ifdef GT_CUDACC
```
Comment (havogt) [inline include/gridtools/common/cuda_util.hpp:77]: cleanup

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftest

File: include/gridtools/fn/run.hpp [hunk 3]
```diff
@@ -26,10 +26,10 @@ namespace gridtools::fn {
 
         template <class Backend, class StageSpecs, class MakeIterator, class Domain, class Sids>
         void run_stencil_stages(
-            Backend, StageSpecs, MakeIterator const &make_iterator, Domain const &domain, Sids &&sids) {
+            Backend backend, StageSpecs, MakeIterator const &make_iterator, Domain const &domain, Sids &&sids) {
```
Comment (fthaler) [inline include/gridtools/fn/run.hpp:29]:
```suggestion
            Backend const &backend, StageSpecs, MakeIterator const &make_iterator, Domain const &domain, Sids &&sids) {
```

File: include/gridtools/fn/run.hpp [hunk 4]
```diff
@@ -40,7 +40,7 @@ namespace gridtools::fn {
             class Vertical,
             class Sids,
             class Seeds>
-        void run_column_stages(Backend,
+        void run_column_stages(Backend backend,
```
Comment (fthaler) [inline include/gridtools/fn/run.hpp:43]:
```suggestion
        void run_column_stages(Backend const &backend,
```

File: include/gridtools/fn/cartesian.hpp [hunk 5]
```diff
@@ -72,28 +72,29 @@ namespace gridtools::fn {
...
                     return make_vertical_executor<Vertical>(
-                        Backend(), m_domain.m_sizes, m_domain.m_offsets, make_iterator());
+                        m_backend, m_domain.m_sizes, m_domain.m_offsets, make_iterator());
                 };
             }
         };
 
         template <class Backend, class Sizes, class Offsets>
-        auto make_backend(Backend, cartesian_domain<Sizes, Offsets> const &d) {
+        auto make_backend(Backend b, cartesian_domain<Sizes, Offsets> const &d) {
```
Comment (fthaler) [inline include/gridtools/fn/cartesian.hpp:95]:
```suggestion
        auto make_backend(Backend const &b, cartesian_domain<Sizes, Offsets> const &d) {
```

File: include/gridtools/fn/executor.hpp [hunk 6]
```diff
@@ -114,16 +122,18 @@ namespace gridtools::fn {
...
+                backend, sizes, offsets, make_iterator};
             return stencil_executor<decltype(data)>{std::move(data)};
         }
 
         // ArgOffset allows passing some args for backend usage while keeping them hidden from the user
         template <class Vertical, int ArgOffset = 0, class Backend, class Sizes, class Offsets, class MakeIterator>
         auto make_vertical_executor(
-            Backend, Sizes const &sizes, Offsets const &offsets, MakeIterator const &make_iterator) {
-            executor_data<Backend, ArgOffset, Sizes, Offsets, MakeIterator> data{sizes, offsets, make_iterator};
+            Backend backend, Sizes const &sizes, Offsets const &offsets, MakeIterator const &make_iterator) {
```
Comment (fthaler) [inline include/gridtools/fn/executor.hpp:134]:
```suggestion
            Backend const &backend, Sizes const &sizes, Offsets const &offsets, MakeIterator const &make_iterator) {
```

File: include/gridtools/fn/unstructured.hpp [hunk 7]
```diff
@@ -154,17 +155,17 @@ namespace gridtools::fn {
...
-                        Backend(), m_domain.m_sizes, m_domain.m_offsets, make_iterator(m_domain.without_offsets()))
+                        m_backend, m_domain.m_sizes, m_domain.m_offsets, make_iterator(m_domain.without_offsets()))
                         .arg(index); // the horizontal index is passed as the first argument
                 };
             }
         };
 
         template <class Backend, class Tables, class Sizes, class Offsets>
-        auto make_backend(Backend, domain_with_offsets<Tables, Sizes, Offsets> const &d) {
+        auto make_backend(Backend b, domain_with_offsets<Tables, Sizes, Offsets> const &d) {
```
Comment (fthaler) [inline include/gridtools/fn/unstructured.hpp:165]:
```suggestion
        auto make_backend(Backend const &b, domain_with_offsets<Tables, Sizes, Offsets> const &d) {
```

**Review by fthaler — CHANGES_REQUESTED** [review]
Looks good. I just propose to be a bit more consistent with passing the backend by const ref where possible to avoid unnecessary copies in case we might have backends with heavier run-time payload in the future.

File: include/gridtools/fn/backend/gpu.hpp [hunk 8]
```diff
@@ -140,14 +141,15 @@ namespace gridtools::fn::backend {
 
         template <class BlockSizes, class Sizes, class StencilStage, class MakeIterator, class Composite>
         void apply_stencil_stage(
-            gpu<BlockSizes>, Sizes const &sizes, StencilStage, MakeIterator make_iterator, Composite &&composite) {
+            gpu<BlockSizes> g, Sizes const &sizes, StencilStage, MakeIterator make_iterator, Composite &&composite) {
```
Comment (havogt) [inline include/gridtools/fn/backend/gpu.hpp:144]:
```suggestion
            gpu<BlockSizes>const &g, Sizes const &sizes, StencilStage, MakeIterator make_iterator, Composite &&composite) {
```

File: include/gridtools/fn/backend/gpu.hpp [hunk 9]
```diff
@@ -178,7 +180,7 @@ namespace gridtools::fn::backend {
             class Composite,
             class Vertical,
             class Seed>
-        void apply_column_stage(gpu<BlockSizes>,
+        void apply_column_stage(gpu<BlockSizes> g,
```
Comment (havogt) [inline include/gridtools/fn/backend/gpu.hpp:183]:
```suggestion
        void apply_column_stage(gpu<BlockSizes> const &g,
```

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch perftest

**Pull request merged**